deploy/runtools/run_farm_deploy_managers.py

```diff
@@ -656,6 +661,37 @@ def clear_fpgas(self) -> None:
                 with StreamLogger('stdout'), StreamLogger('stderr'):
                     run(f"xbutil reset -d {card_bdf} --force")
 
+    def localize_xclbin(self, slotno: int) -> None:
+        """ download xclbin URI to remote node. """
+        assert slotno < len(self.parent_node.sim_slots)
+        serv = self.parent_node.sim_slots[slotno]
+        hwcfg = serv.get_resolved_server_hardware_config()
+        if re.match(_RFC_3986_PATTERN, hwcfg.xclbin):
+            remote_home_dir = self.parent_node.get_sim_dir()
+            remote_sim_dir = """{}/sim_slot_{}/""".format(remote_home_dir, slotno)
+            hwcfg.local_xclbin = './local.xclbin'
+
+            def get_xclbin(xclbin_uri:str, xclbin_lpath:PathLike) -> None:
+                # TODO instead of blowing up if the file exists, consider using fsspec
+                # filecache https://filesystem-spec.readthedocs.io/en/latest/features.html#caching-files-locally
+                # so that multiple slots using the same xclbin only grab it once and
+                # we only download it if it has changed at the source.
+                # HOWEVER, 'filecache' isn't thread/process safe and I'm not sure whether
+                # this runs in @parallel for fabric
+                lpath = Path(xclbin_lpath)
```

Comment:
I hate to say this... but we should probably port to using `pathlib` everywhere. (just a comment)

Reply:
Yeah, it might make sense.  I've started using it only because it makes `Path` methods available to tab-completion in whatever IDE I'm using and I'm too lazy to remember which functions need to be imported from `os.path` to do whatever I'm doing.

Note that there are some significant wrinkles with regards to `pathlib` and `fsspec` interoperability (see https://github.com/fsspec/filesystem_spec/issues/539).  Enough so that folks at QuantSight have developed https://github.com/fsspec/universal_pathlib as an adapter for `pathlib` usage with `fsspec`

Reply:
Should we use `universal_pathlib` instead of `pathlib` everywhere then? Based on this it seems like it's a superset of functionality? https://github.com/fsspec/universal_pathlib/blob/main/notebooks/examples.ipynb.

I'm slightly concerned about future bugs when using `Path`s (or `UPath`s) with Fabric since `Path` objects are relative to localhost (i.e. `Path("/remote/path").mkdir()` will create a local directory when you meant to do this with Fabric). In any case, this discussion should happen in another PR.

Reply:
I agree that we should pull this out into another PR.  I captured #1148 for discussion.